drivers/vdpa/vfe/vfe_vdpa.c

```diff
@@ -0,0 +1,1196 @@
+/*				   (1ULL << VHOST_USER_PROTOCOL_F_LOG_SHMFD) | \
+				 (1ULL << VHOST_USER_PROTOCOL_F_MQ) | \
+				 (1ULL << VHOST_USER_PROTOCOL_F_NET_MTU) | \
+				 (1ULL << VHOST_USER_PROTOCOL_F_STATUS))
+*/
+
+TAILQ_HEAD(vfe_vdpa_privs, vfe_vdpa_priv) vfe_priv_list =
+						  TAILQ_HEAD_INITIALIZER(vfe_priv_list);
+static pthread_mutex_t priv_list_lock = PTHREAD_MUTEX_INITIALIZER;
+
+static struct vfe_vdpa_priv *
+vfe_vdpa_find_priv_resource_by_vdev(struct rte_vdpa_device *vdev)
+{
+	struct vfe_vdpa_priv *priv;
+	int found = 0;
+
+	pthread_mutex_lock(&priv_list_lock);
+	TAILQ_FOREACH(priv, &vfe_priv_list, next) {
+		if (vdev == priv->vdev) {
+			found = 1;
+			break;
+		}
+	}
+	pthread_mutex_unlock(&priv_list_lock);
+	if (!found) {
+		DRV_LOG(ERR, "Invalid vDPA device: %s.", vdev->device->name);
+		rte_errno = EINVAL;
+		return NULL;
+	}
+	return priv;
+}
+
+static uint16_t
+vfe_get_nr_vq(struct virtio_hw *hw)
+{
+	uint16_t nr_vq = hw->max_queue_pairs * 2;
+
+	if (virtio_with_feature(hw, VIRTIO_NET_F_MQ) ||
+			virtio_with_feature(hw, VIRTIO_NET_F_RSS)) {
+		vfe_virtio_read_dev_config(hw,
+			offsetof(struct virtio_net_config, max_virtqueue_pairs),
+			&hw->max_queue_pairs,
+			sizeof(hw->max_queue_pairs));
+	} else {
+		DRV_LOG(DEBUG,
+				 "Neither VIRTIO_NET_F_MQ nor VIRTIO_NET_F_RSS are supported");
+		hw->max_queue_pairs = 1;
+	}
+
+	nr_vq = hw->max_queue_pairs * 2;
+	if (virtio_with_feature(hw, VIRTIO_NET_F_CTRL_VQ))
+		nr_vq += 1;
+
+	DRV_LOG(DEBUG,"virtio nr_vq is %d",nr_vq);
+
+	return nr_vq;
+}
+
+static int
+vfe_vdpa_get_queue_num(struct rte_vdpa_device *vdev, uint32_t *queue_num)
+{
+	struct vfe_vdpa_priv *priv =
+		vfe_vdpa_find_priv_resource_by_vdev(vdev);
+
+	if (priv == NULL) {
+		DRV_LOG(ERR, "Invalid vDPA device: %s.", vdev->device->name);
+		return -1;
+	}
+	// TO_DO:for net, should use max_virtqueue_pairs? may be is same
+	*queue_num = priv->hw_nr_virtqs;
+	DRV_LOG(DEBUG, "vid %d queue num is %d", priv->vid, *queue_num);
+	return 0;
+}
+
+static int
+vfe_vdpa_get_vdpa_features(struct rte_vdpa_device *vdev, uint64_t *features)
+{
+	struct vfe_vdpa_priv *priv =
+		vfe_vdpa_find_priv_resource_by_vdev(vdev);
+	struct virtio_hw *hw = NULL;
+
+	if (priv == NULL) {
+		DRV_LOG(ERR, "Invalid vDPA device: %s.", vdev->device->name);
+		return -1;
+	}
+	hw = &priv->vpdev.hw;
+	*features = hw->guest_features;
+	*features |= (1ULL << VHOST_USER_F_PROTOCOL_FEATURES);
+   
+	return 0;
+}
+
+static int
+vfe_vdpa_get_protocol_features(struct rte_vdpa_device *vdev,
+		uint64_t *features)
+{
+	struct vfe_vdpa_priv *priv =
+		vfe_vdpa_find_priv_resource_by_vdev(vdev);
+
+	if (priv == NULL) {
+		DRV_LOG(ERR, "Invalid vDPA device: %s.", vdev->device->name);
+		return -1;
+	}
+	*features = VFE_VDPA_PROTOCOL_FEATURES;
+	return 0;
+}
+
+static uint64_t
+hva_to_gpa(int vid, uint64_t hva)
+{
+	struct rte_vhost_memory *mem = NULL;
+	struct rte_vhost_mem_region *reg;
+	uint32_t i;
+	uint64_t gpa = 0;
+
+	if (rte_vhost_get_mem_table(vid, &mem) < 0)
+		goto exit;
```

Comment:
exit label frees the memory.
there is no point in freeing returning memory pointer when get_mem_table() returns error.

if rte_vhost_get_mem_table() fails, return directly from there with return 0.
also add error print because this must not happen.